pkg/atomix/log/service.go

```diff
@@ -0,0 +1,481 @@
+}
+
+// init initializes the log service
+func (m *Service) init() {
+	m.Executor.RegisterUnaryOperation(opPut, m.Put)
+	m.Executor.RegisterUnaryOperation(opRemove, m.Remove)
+	m.Executor.RegisterUnaryOperation(opGet, m.Get)
+	m.Executor.RegisterUnaryOperation(opFirstEntry, m.FirstEntry)
+	m.Executor.RegisterUnaryOperation(opLastEntry, m.LastEntry)
+	m.Executor.RegisterUnaryOperation(opPrevEntry, m.PrevEntry)
+	m.Executor.RegisterUnaryOperation(opNextEntry, m.NextEntry)
+	m.Executor.RegisterUnaryOperation(opExists, m.Exists)
+	m.Executor.RegisterUnaryOperation(opSize, m.Size)
+	m.Executor.RegisterUnaryOperation(opClear, m.Clear)
+}
+
+// LinkedEntryValue is a doubly linked LogEntryValue
+type LinkedLogEntryValue struct {
+	*LogEntryValue
+	Prev *LinkedLogEntryValue
+	Next *LinkedLogEntryValue
+}
+
+// Snapshot takes a snapshot of the service
+func (m *Service) Snapshot(writer io.Writer) error {
+	if err := m.SessionizedService.Snapshot(writer); err != nil {
+		return err
+	}
+
+	if err := util.WriteVarUint64(writer, m.lastIndex); err != nil {
+		return err
+	}
+
+	entry := m.firstEntry
+	for entry != nil {
+		if err := util.WriteValue(writer, entry.LogEntryValue, proto.Marshal); err != nil {
+			return err
+		}
+		entry = entry.Next
+	}
+	return nil
+}
+
+// Install restores the service from a snapshot
+func (m *Service) Install(reader io.Reader) error {
+	if err := m.SessionizedService.Install(reader); err != nil {
+		return err
+	}
+
+	lastIndex, err := util.ReadVarUint64(reader)
+	if err != nil {
+		return err
+	}
+	m.lastIndex = lastIndex
+
+	entryCount, err := util.ReadVarInt(reader)
+	if err != nil {
+		return err
+	}
+
+	var prevEntry *LinkedLogEntryValue
+	m.firstEntry = nil
+	m.lastEntry = nil
+	m.indexes = make(map[uint64]*LinkedLogEntryValue)
+	for i := 0; i < entryCount; i++ {
+		value, err := util.ReadValue(reader, func(data []byte) (*LogEntryValue, error) {
+			entry := &LogEntryValue{}
+			if err := proto.Unmarshal(data, entry); err != nil {
+				return nil, err
+			}
+			return entry, nil
+		})
+		if err != nil {
+			return err
+		}
+
+		linkedEntry := &LinkedLogEntryValue{
+			LogEntryValue: value.(*LogEntryValue),
+		}
+		if m.firstEntry == nil {
+			m.firstEntry = linkedEntry
+		}
+
+		m.indexes[linkedEntry.Index] = linkedEntry
+		if prevEntry != nil {
+			prevEntry.Next = linkedEntry
+			linkedEntry.Prev = prevEntry
+		}
+		prevEntry = linkedEntry
+		m.lastEntry = linkedEntry
+	}
+	return nil
+}
+
+// Put puts a key/value pair in the log
+func (m *Service) Put(value []byte) ([]byte, error) {
+	request := &PutRequest{}
+	if err := proto.Unmarshal(value, request); err != nil {
+		return nil, err
+	}
+
+	var oldEntry *LinkedLogEntryValue
+	if request.Index > 0 {
+		oldEntry = m.indexes[request.Index]
+		if oldEntry != nil && oldEntry.Index != request.Index {
+			return proto.Marshal(&PutResponse{
+				Status: UpdateStatus_PRECONDITION_FAILED,
+			})
+		}
+	} else {
+		oldEntry = m.indexes[m.lastIndex]
```

Comment:
This is allowing an updated, which we should disallow with a `PRECONDITION_FAILED` response. Essentially, very write is conditioned on the entry not existing.